// @tencent-ai/agent-sdk/lib/query.js
"use strict";
/**
 * Query Implementation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = exports.query = void 0;
const query_controller_1 = require("./_internal/query-controller");
const connect_1 = require("./connect");
const errors_1 = require("./errors");
const transport_1 = require("./transport");
const utils_1 = require("./utils");
/**
 * Create a query for interacting with the agent.
 *
 * When `prompt` is omitted and `options.resume` is provided, the query will
 * return historical messages from the resumed session without triggering
 * a new agent response.
 */
function query(params) {
    return new Query(params.prompt, params.options);
}
exports.query = query;
/**
 * Query class - an AsyncGenerator that yields messages.
 */
class Query {
    constructor(prompt, options) {
        // Note: Session lock is now acquired in initialize() to:
        // 1. Reduce lock leak risk (only acquire when actually used)
        // 2. Allow fork operations which don't need locks
        var _a, _b, _c;
        this.prompt = prompt;
        this.options = options;
        this.initialized = false;
        this.iterator = null;
        this.connectPromise = null;
        // Session lock tracking
        this.lockedSessionId = null;
        // Session ID captured from CLI messages
        this._sessionId = null;
        // Validate sessionId format
        if (options === null || options === void 0 ? void 0 : options.sessionId) {
            const sessionIdPattern = /^[a-zA-Z0-9][a-zA-Z0-9\-_:]*$/;
            if (!sessionIdPattern.test(options.sessionId)) {
                throw new Error(`Invalid session ID format: "${options.sessionId}". ` +
                    'Session IDs support numbers, letters, hyphens, underscores, and colons, ' +
                    'but must start with a letter or number.');
            }
        }
        // Parse systemPrompt options
        const systemPrompt = typeof (options === null || options === void 0 ? void 0 : options.systemPrompt) === 'string'
            ? options.systemPrompt
            : undefined;
        const appendSystemPrompt = typeof (options === null || options === void 0 ? void 0 : options.systemPrompt) === 'object'
            ? options.systemPrompt.append
            : undefined;
        // Build transport options (transport handles SDK MCP server extraction)
        const transportOptions = {
            executablePath: options === null || options === void 0 ? void 0 : options.pathToCodebuddyCode,
            executable: options === null || options === void 0 ? void 0 : options.executable,
            executableArgs: options === null || options === void 0 ? void 0 : options.executableArgs,
            cwd: options === null || options === void 0 ? void 0 : options.cwd,
            env: options === null || options === void 0 ? void 0 : options.env,
            permissionMode: options === null || options === void 0 ? void 0 : options.permissionMode,
            // Model options
            model: options === null || options === void 0 ? void 0 : options.model,
            fallbackModel: options === null || options === void 0 ? void 0 : options.fallbackModel,
            maxTurns: options === null || options === void 0 ? void 0 : options.maxTurns,
            // Tool options
            allowedTools: options === null || options === void 0 ? void 0 : options.allowedTools,
            disallowedTools: options === null || options === void 0 ? void 0 : options.disallowedTools,
            tools: options === null || options === void 0 ? void 0 : options.tools,
            // Session options
            sessionId: options === null || options === void 0 ? void 0 : options.sessionId,
            continue: options === null || options === void 0 ? void 0 : options.continue,
            resume: options === null || options === void 0 ? void 0 : options.resume,
            forkSession: options === null || options === void 0 ? void 0 : options.forkSession,
            // MCP options (transport extracts SDK servers automatically)
            mcpServers: options === null || options === void 0 ? void 0 : options.mcpServers,
            strictMcpConfig: options === null || options === void 0 ? void 0 : options.strictMcpConfig,
            // Settings
            settingSources: options === null || options === void 0 ? void 0 : options.settingSources,
            additionalDirectories: options === null || options === void 0 ? void 0 : options.additionalDirectories,
            // Output options
            includePartialMessages: options === null || options === void 0 ? void 0 : options.includePartialMessages,
            // System prompt options
            systemPrompt,
            appendSystemPrompt,
            // Thinking & effort
            maxThinkingTokens: options === null || options === void 0 ? void 0 : options.maxThinkingTokens,
            thinking: options === null || options === void 0 ? void 0 : options.thinking,
            effort: options === null || options === void 0 ? void 0 : options.effort,
            // Extra args
            extraArgs: options === null || options === void 0 ? void 0 : options.extraArgs,
            stderr: options === null || options === void 0 ? void 0 : options.stderr,
            // Transport control request timeout
            requestTimeoutMs: options === null || options === void 0 ? void 0 : options.requestTimeoutMs,
        };
        this.transport = (0, transport_1.createTransport)(transportOptions);
        this.abortController = (_a = options === null || options === void 0 ? void 0 : options.abortController) !== null && _a !== void 0 ? _a : new AbortController();
        // QueryController owns initialize + hook registry + control-request routing.
        // The canUseTool getter is lazy because options shouldn't be re-read on every
        // call (they don't change for query()), but it matches Session's pattern.
        this.controller = new query_controller_1.QueryController(this.transport, options === null || options === void 0 ? void 0 : options.hooks, () => options === null || options === void 0 ? void 0 : options.canUseTool, () => this.abortController.signal, () => options === null || options === void 0 ? void 0 : options.elicitation);
        // Register control request handler on transport so control requests
        // are dispatched directly in handleLine(), bypassing the message queue.
        this.transport.onControlRequest(request => {
            // eslint-disable-next-line no-void
            void this.controller.handleControlRequest(request).catch(() => {
                // Individual control request errors shouldn't affect the query
            });
        });
        // elicitation/complete 收口通知（fire-and-forget，无需订阅握手）：
        // agent 侧查单证实/超时后撤卡/翻终态。
        if ((_b = options === null || options === void 0 ? void 0 : options.elicitation) === null || _b === void 0 ? void 0 : _b.complete) {
            this.transport.onNotification('elicitation', data => {
                var _a, _b;
                // eslint-disable-next-line no-void
                void Promise.resolve((_b = (_a = options.elicitation) === null || _a === void 0 ? void 0 : _a.complete) === null || _b === void 0 ? void 0 : _b.call(_a, data.data)).catch(() => { });
            });
        }
        // Auto-connect: start CLI process immediately (non-blocking)
        // This reduces latency for the first message
        this.connectPromise = this.transport.connect().catch(() => {
            // Ignore connection errors - they'll be reported when iteration starts
        });
        // Track initial values (passed to CLI via command line args)
        this._initialPermissionMode = (_c = options === null || options === void 0 ? void 0 : options.permissionMode) !== null && _c !== void 0 ? _c : 'default';
        this._initialModel = options === null || options === void 0 ? void 0 : options.model;
        // Current state starts same as initial
        this._currentPermissionMode = this._initialPermissionMode;
        this._currentModel = this._initialModel;
    }
    // ============= AsyncGenerator Implementation =============
    [Symbol.asyncIterator]() {
        if (!this.iterator) {
            this.iterator = this.createIterator();
        }
        return this.iterator;
    }
    async next() {
        return this[Symbol.asyncIterator]().next();
    }
    async return() {
        await this.interrupt();
        return { done: true, value: undefined };
    }
    async throw(e) {
        await this.interrupt();
        throw e;
    }
    // ============= Connection Methods =============
    /**
     * Explicitly connect and initialize the query.
     * This is called automatically when iterating, but can be called
     * manually if you want to ensure the CLI is ready before iterating.
     *
     * Note: The constructor already triggers auto-connection, so this
     * method is mainly useful if you want to await the connection completion.
     *
     * @example
     * ```typescript
     * const q = query({ prompt: 'Hello' });
     * await q.connect();  // Wait for CLI process to be ready
     * for await (const msg of q) {
     *   // First message arrives faster
     * }
     * ```
     */
    async connect() {
        await this.transport.connect();
        await this.initialize();
    }
    // ============= Control Methods =============
    /**
     * Get the session ID to use for control requests.
     * Prefers the session_id captured from CLI messages, falls back to options.
     */
    getSessionId() {
        var _a, _b;
        return this._sessionId || ((_a = this.options) === null || _a === void 0 ? void 0 : _a.sessionId) || ((_b = this.options) === null || _b === void 0 ? void 0 : _b.resume) || undefined;
    }
    async interrupt() {
        try {
            await this.transport.sendControlRequest({
                subtype: 'interrupt',
                session_id: this.getSessionId(),
                reason: 'User interrupt',
            });
        }
        catch (_a) {
            // Ignore errors during interrupt
        }
        // Don't close transport - let the iterator continue to receive
        // the result message and close naturally
    }
    /**
     * Set the permission mode for the query.
     * - Before iteration starts: only updates local state
     * - After iteration starts: sends fire-and-forget control request
     */
    async setPermissionMode(mode) {
        this._currentPermissionMode = mode;
        // Only sync to CLI if session is established (has session ID)
        if (this._sessionId) {
            this.transport.sendControlRequest({
                subtype: 'set_permission_mode',
                session_id: this._sessionId,
                mode,
            }).catch(() => { });
        }
    }
    /**
     * Get the current permission mode.
     */
    getPermissionMode() {
        return this._currentPermissionMode;
    }
    /**
     * Set the model for the query.
     * - Before iteration starts: only updates local state
     * - After iteration starts: sends fire-and-forget control request
     */
    async setModel(model) {
        this._currentModel = model;
        // Only sync to CLI if session is established (has session ID)
        if (this._sessionId) {
            this.transport.sendControlRequest({
                subtype: 'set_model',
                session_id: this._sessionId,
                model,
            }).catch(() => { });
        }
    }
    /**
     * Get the current model.
     */
    getModel() {
        return this._currentModel;
    }
    async setMaxThinkingTokens(maxThinkingTokens) {
        await this.transport.sendControlRequest({
            subtype: 'set_max_thinking_tokens',
            session_id: this.getSessionId(),
            max_thinking_tokens: maxThinkingTokens,
        });
    }
    // ============= Query Methods =============
    async supportedCommands() {
        var _a;
        const response = await this.transport.sendControlRequest({
            subtype: 'initialize',
        });
        return (_a = response.commands) !== null && _a !== void 0 ? _a : [];
    }
    async supportedModels() {
        var _a;
        const response = await this.transport.sendControlRequest({
            subtype: 'initialize',
        });
        return (_a = response.models) !== null && _a !== void 0 ? _a : [];
    }
    async mcpServerStatus() {
        var _a;
        const response = await this.transport.sendControlRequest({
            subtype: 'mcp_status',
        });
        return (_a = response.mcp_servers) !== null && _a !== void 0 ? _a : [];
    }
    async accountInfo() {
        var _a;
        const response = await this.transport.sendControlRequest({
            subtype: 'initialize',
        });
        return (_a = response.account) !== null && _a !== void 0 ? _a : {};
    }
    async streamInput(stream) {
        for await (const message of stream) {
            this.transport.sendUserMessage(message);
        }
    }
    // ============= Private Methods =============
    async *createIterator() {
        var _a;
        try {
            // Wait for connection to complete (if in progress)
            if (this.connectPromise) {
                await this.connectPromise;
            }
            await this.initialize();
            // Send prompt (CLI decides whether to replay history based on hasPrompt in initialize)
            await this.sendPrompt();
            // Note: Control requests (can_use_tool, hook_callback) are handled
            // directly by the transport via onControlRequest() and never appear here.
            for await (const message of this.transport.messages()) {
                if (this.abortController.signal.aborted) {
                    break;
                }
                if ((0, utils_1.isMessage)(message)) {
                    // Capture session_id from CLI messages for use in control requests
                    if (!this._sessionId && 'session_id' in message && message.session_id) {
                        this._sessionId = message.session_id;
                        // Send pending permission mode change (if set before iteration)
                        if (this._currentPermissionMode !== this._initialPermissionMode) {
                            this.transport.sendControlRequest({
                                subtype: 'set_permission_mode',
                                session_id: this._sessionId,
                                mode: this._currentPermissionMode,
                            }).catch(() => { });
                        }
                        // Send pending model change (if set before iteration)
                        if (this._currentModel !== this._initialModel) {
                            this.transport.sendControlRequest({
                                subtype: 'set_model',
                                session_id: this._sessionId,
                                model: this._currentModel,
                            }).catch(() => { });
                        }
                    }
                    if (message.type === 'stream_event' &&
                        !((_a = this.options) === null || _a === void 0 ? void 0 : _a.includePartialMessages)) {
                        continue;
                    }
                    // Check for execution error BEFORE yielding
                    // This ensures the exception is thrown even if caller breaks after receiving result
                    if (message.type === 'result') {
                        if (message.is_error && 'errors' in message && message.errors && message.errors.length > 0) {
                            throw new errors_1.ExecutionError(message.errors, message.subtype);
                        }
                        yield message;
                        break;
                    }
                    yield message;
                }
            }
        }
        finally {
            this.cleanup();
        }
    }
    async initialize() {
        var _a, _b, _c, _d, _e, _f;
        if (this.initialized) {
            return;
        }
        // Delayed lock acquisition: acquire lock when actually used to reduce leak risk
        // Fork operations don't need locks since they create independent copies
        if (((_a = this.options) === null || _a === void 0 ? void 0 : _a.resume) && !((_b = this.options) === null || _b === void 0 ? void 0 : _b.forkSession)) {
            if (!(0, connect_1.acquireSessionLock)(this.options.resume)) {
                throw new Error(`Session ${this.options.resume} is already in use`);
            }
            this.lockedSessionId = this.options.resume;
        }
        const response = await this.controller.initialize({
            agents: (_c = this.options) === null || _c === void 0 ? void 0 : _c.agents,
            systemPrompt: (_d = this.options) === null || _d === void 0 ? void 0 : _d.systemPrompt,
            environment: (_e = this.options) === null || _e === void 0 ? void 0 : _e.environment,
            endpoint: (_f = this.options) === null || _f === void 0 ? void 0 : _f.endpoint,
            // Tell CLI whether prompt will follow (controls resume history replay).
            // hasPrompt=true means CLI should NOT replay history.
            hasPrompt: this.prompt !== undefined && this.prompt !== '',
        });
        this.initialized = true;
        // When ``options.model`` was unset, sync the CLI-reported model into both
        // current *and* initial — otherwise the pending-change check at iteration
        // start would treat the CLI's reported value as a user-initiated change.
        if (!this._currentModel && response.currentModelId) {
            this._currentModel = response.currentModelId;
            this._initialModel = response.currentModelId;
        }
    }
    async sendPrompt() {
        var _a, _b, _c, _d;
        // If no prompt is provided (e.g., resume mode to get history only),
        // don't send any message. CLI will return history when no prompt is sent.
        if (this.prompt === undefined) {
            return;
        }
        if (typeof this.prompt === 'string') {
            // Empty string also triggers history-only mode
            if (this.prompt.trim() === '') {
                return;
            }
            this.transport.sendUserMessage(this.prompt, (_a = this.options) === null || _a === void 0 ? void 0 : _a.traceId, (_b = this.options) === null || _b === void 0 ? void 0 : _b.parentSpanId);
        }
        else {
            for await (const message of this.prompt) {
                if (!message.trace_id && ((_c = this.options) === null || _c === void 0 ? void 0 : _c.traceId)) {
                    message.trace_id = this.options.traceId;
                }
                if (!message.parent_span_id && ((_d = this.options) === null || _d === void 0 ? void 0 : _d.parentSpanId)) {
                    message.parent_span_id = this.options.parentSpanId;
                }
                this.transport.sendUserMessage(message);
            }
        }
    }
    /**
     * Cleanup all resources.
     */
    cleanup() {
        // Release session lock if held
        if (this.lockedSessionId) {
            (0, connect_1.releaseSessionLock)(this.lockedSessionId);
            this.lockedSessionId = null;
        }
        // Close main transport (handles SDK MCP cleanup internally)
        this.transport.close();
    }
}
exports.Query = Query;
